src/app/modules/main/communities/module.nim

```diff
@@ -531,13 +531,71 @@ method callbackFromAuthentication*(self: Module, authenticated: bool) =
 method getCommunityPublicKeyFromPrivateKey*(self: Module, communityPrivateKey: string): string =
   result = self.controller.getCommunityPublicKeyFromPrivateKey(communityPrivateKey)
 
+method updateTokenModelForCommunity*(self: Module, communityId: string, sharedAddresses: seq[string]) =
+  self.controller.asyncCheckPermissionsToJoin(communityId, sharedAddresses)
+  self.controller.asyncCheckAllChannelsPermissions(communityId, sharedAddresses)
+
 method prepareTokenModelForCommunity*(self: Module, communityId: string) =
   let community = self.controller.getCommunityById(communityId)
   var tokenPermissionsItems: seq[TokenPermissionItem] = @[]
 
   for id, tokenPermission in community.tokenPermissions:
-    let chats = self.controller.getChatDetailsByIds(tokenPermission.chatIDs)
+    let chats = community.getCommunityChats(tokenPermission.chatIds)
     let tokenPermissionItem = buildTokenPermissionItem(tokenPermission, chats)
     tokenPermissionsItems.add(tokenPermissionItem)
 
-  self.view.spectatedCommunityPermissionModel.setItems(tokenPermissionsItems)
\ No newline at end of file
+  self.view.spectatedCommunityPermissionModel.setItems(tokenPermissionsItems)
+  self.updateTokenModelForCommunity(communityId, @[])
+
+proc applyPermissionResponse*(self: Module, communityId: string, permissions: Table[string, CheckPermissionsResultDto]) =
+  let community = self.controller.getCommunityById(communityId)
+  for id, criteriaResult in permissions:
+    if community.tokenPermissions.hasKey(id):
+      let tokenPermissionItem = self.view.spectatedCommunityPermissionModel.getItemById(id)
+      if tokenPermissionItem.id == "":
+        continue
```

Comment:
```suggestion
      if tokenPermissionItem.id == "":
        warn "no permission in model", id
        continue
```

`spectatedCommunityPermissionModel` should already contain all permissions from response, right?

Reply:
It should yes